### ISSUE TITLE
Many CPU cores and scheduler request delays do not play well together for backup projects

### ISSUE THREAD

**Issue by sirzooro** [BOINC/boinc#1941]
When BOINC Client cannot download new WUs from projects with non-zero resource share, it tries to download them from projects with resource share set to 0. However when it does this, it requests only minimum required number of tasks to start them immediately and does not try to build any buffer. When host has many CPU cores and backup project has minimum scheduler request delay set to few minutes, host is not able to immediately download new WUs as necessary, it has to wait until this request delay time ends. Please change this. In such case BOINC Client should at least work as if work buffer size was set to zero. You can also consider using configured value or some fraction of it, but this sometimes may not be desirable - BOINC may download a lot of work from backup project, when them main one will stop providing work during maintenance period.

**Comment by ChristianBeer** [comment]
The reason why only minimal work is requested from the backup project is exactly as you stated. If we would build a cache we could download too much work from the backup project which is not what it should do. In your case it seems that the backup project you chose has work that is not suitable to be a backup project. I guess it has short tasks that finish before the request delay is over.

In my opinion this should not be changed in the Client. I don't see how setting work buffer to 0 temporarily will help the situation. I'm keeping this open so others can voice there opinions.

**Comment by sirzooro** [comment]
When buffer is set to 0, BOINC Client downloads new tasks few minutes before previous ones end - at least it works this way for PrimeGrid. This creates small buffer which is enough to avoid problem described above.

**Comment by ChristianBeer** [comment]
This only works if the runtime estimates of the backup project are reasonably accurate, if they are not you will end up with a cache that is too large. It's better to be conservative here and maybe loose some CPUc cycles. My guess is that there is an equal number of people out there that would protest this change and would expect a backup project to not have a cache ever.

**Comment by sirzooro** [comment]
If there are two opposite opinions on this topic, it would be good to go with 3rd one - provide end user a way to configure this. New option in cc_config.xml to configure size of backup buffer would be fine.

**Comment by Aurum420** [comment]
I wish I knew the definition of a BOINC backup project. I've never seen it in BOINC.
I've also never seen this behavior: "When BOINC Client cannot download new WUs from projects with non-zero resource share, it tries to download them from projects with resource share set to 0." 
Sounds like a server-side configuration issue for a specific project, maybe?

**Comment by sirzooro** [comment]
@Aurum420 this is documented here: https://boinc.berkeley.edu/wiki/Preferences#List_of_project_preferences

**Comment by Aurum420** [comment]
Thanks, let me rephrase. I've only ever seen one project that behaves as stated and that's niverse. The others seem to ignore 0.

**Comment by sirzooro** [comment]
This parameter is used bu Boinc Manager. Project servers only store it and send to manager. I use it with PrimeGrid - in fact in this project it is recommended to increase change to be first discoverer of new prime. I also tried to use this with about 2 other project but there were some issues with downloading WUs - one of them ran out of WUs for some time, and another one required long delay (something like 1 minute) between every connection to server.
